kernel/src/syscall.rs

```diff
@@ -1,21 +1,108 @@
-//! Tock syscall number definitions.
+//! Tock syscall number definitions and arch-agnostic interface trait.
+
+use process;
 
 /// The syscall number assignments.
 #[derive(Copy, Clone, Debug)]
-crate enum Syscall {
+pub enum Syscall {
     /// Return to the kernel to allow other processes to execute or to wait for
     /// interrupts and callbacks.
-    YIELD = 0,
+    ///
+    /// SVC_NUM = 0
+    YIELD,
 
     /// Pass a callback function to the kernel.
-    SUBSCRIBE = 1,
+    ///
+    /// SVC_NUM = 1
+    SUBSCRIBE {
+        driver_number: usize,
+        subdriver_number: usize,
+        callback_ptr: *mut (),
+        appdata: usize,
+    },
 
     /// Instruct the kernel or a capsule to perform an operation.
-    COMMAND = 2,
+    ///
+    /// SVC_NUM = 2
+    COMMAND {
+        driver_number: usize,
+        subdriver_number: usize,
+        arg0: usize,
+        arg1: usize,
+    },
 
     /// Share a memory buffer with the kernel.
-    ALLOW = 3,
+    ///
+    /// SVC_NUM = 3
+    ALLOW {
+        driver_number: usize,
+        subdriver_number: usize,
+        allow_address: *mut u8,
+        allow_size: usize,
+    },
 
     /// Various memory operations.
-    MEMOP = 4,
+    ///
+    /// SVC_NUM = 4
+    MEMOP { operand: usize, arg0: usize },
+}
+
+/// Why the process stopped executing and execution returned to the kernel.
+pub enum ContextSwitchReason {
+    /// Process exceeded its timeslice, otherwise catch-all.
+    Other,
+    /// Process called a syscall.
+    SyscallFired,
+    /// Process triggered the hardfault handler.
+    Fault,
+}
+
+/// This trait must be implemented by the architecture of the chip Tock is
+/// running on. It allows the kernel to manage processes in an
+/// architecture-agnostic manner.
+pub trait SyscallInterface {
+    /// Some architecture-specific struct containing per-process state that must
+    /// be kept while the process is not running. For example, for keeping CPU
+    /// registers that aren't stored on the stack.
+    type StoredState: Default;
+
+    /// Allows the kernel to query to see why the process stopped running. This
+    /// function can only be called once to get the last state of the process
+    /// and why the process context switched back to the kernel.
+    ///
+    /// An implementor of this function is free to reset any state that was
+    /// needed to gather this information when this function is called.
+    unsafe fn get_context_switch_reason(&self) -> ContextSwitchReason;
```

Comment:
I don't know what I want this to be named, but something better. `get_context_switch_reason` doesn't imply that this function seems to also clean up a switched process and can only be called once. Maybe something like `cleanup_switched_process(&self) -> ContextSwitchReason`? Especially because, for comparison, `get_syscall` can be called as many times as you please with no side effects.

Reply:
What about `get_and_reset_context_switch_reason()`?